Sample with Kafka trigger and NServiceBus SendOnly endpoint in Azure functions (#6200)
Particular/docs.particular.net: samples/azure-functions/service-bus-kafka/sample.md

- The ConsoleEndpoint is the starting point of the sample, which produces numerous Kafka events.
- The Azure Function uses a Kafka trigger to consume the events and send NServiceBus messages back to the ConsoleEndpoint via Azure ServiceBus.

For more information about Kafka and NServiceBus read the blogpost [Let's talk about Kafka](https://particular.net/blog/lets-talk-about-kafka).

downloadbutton

## Prerequisites

### Configure Connection string

To use the sample a valid Service Bus connection string must be provided in the `local.settings.json` file in the `AzureFunctions.KafkaTrigger.FunctionsHostBuilder` project and as an environment variable named `AzureServiceBus_ConnectionString`

### Kafka broker

This sample requires Kafka available to store the events being produced.

#### Set up Kafka Docker container

To set up a Docker container with Kafka, the sample contains a `docker-compose.yml` file. From a CLI like PowerShell, using the following command in the solution folder, a Docker image can be downloaded and executed:

```
docker-compose up
```

> NOTE: If you're running Docker in Windows, you might run get an error saying _no matching manifest for windows/amd64 in the manifest list entries_. This can be solved by running the daemon in [experimental mode](https://stackoverflow.com/questions/48066994/docker-no-matching-manifest-for-windows-amd64-in-the-manifest-list-entries).

Author: ```suggestion
NOTE: When running Docker in Windows, it's possible to get an error saying _no matching manifest for windows/amd64 in the manifest list entries_. This can be solved by running the daemon in [experimental mode](https://stackoverflow.com/questions/48066994/docker-no-matching-manifest-for-windows-amd64-in-the-manifest-list-entries).
```

reviewed: 2023-09-13
---

This sample shows how to process Kafka events using an Azure Functions trigger and how to follow up on those events using an NServiceBus SendOnly endpoint in an Azure Function.

### Kafka and NServiceBus

[Kafka](https://kafka.apache.org/) is an event streaming broker, similar to [Azure Event Hubs](https://azure.microsoft.com/en-us/products/event-hubs). Event streaming brokers are used to store massive amounts of events that are coming in, for example from IoT devices. NServiceBus on the other hand works on top of messaging brokers like Azure Service Bus, RabbitMQ and Amazon SQS/SNS. They can complement each other as shown in this sample, which has two projects.

- A ConsoleEndpoint is the starting point of the sample, which products numerous events to Kafka.
- An Azure Function using a Kafka trigger to consume the events and at a certain point sends a message using an NServiceBus SendOnly endpoint back to the ConsoleEndpoint.

For more information about Kafka and NServiceBus read the blogpost [Let's talk about Kafka](https://particular.net/blog/lets-talk-about-kafka).

downloadbutton

## Prerequisites

### Configure Connection string

> To use the sample a valid Service Bus connection string must be provided
> 
> - in the `local.settings.json` file in the `AzureFunctions.KafkaTrigger.FunctionsHostBuilder` project.
> 
> - as an environment variable named `AzureServiceBus_ConnectionString`.

Reviewer: ```suggestion
To use the sample a valid Service Bus connection string must be provided in the `local.settings.json` file in the `AzureFunctions.KafkaTrigger.FunctionsHostBuilder` project and as an environment variable named `AzureServiceBus_ConnectionString`
```

### Kafka broker

This sample requires Kafka available to store the events being produced.

> #### Set up Kafka Docker container

Reviewer: I wonder if we could call docker-compose up in the start up of the console? To improve the F5 experience? With a comment stating don't do this in production :) Could be a separate app in the sln that would do that. I may try to see if I can make this work.
